File: src/main/java/moe/sota/decompiler/controllers/TabsController.java
package moe.sota.decompiler.controllers;

import lombok.Getter;
import moe.sota.decompiler.menus.file.FileCloseTab;
import moe.sota.decompiler.models.FileModel;
import moe.sota.decompiler.services.PreferenceService;
import moe.sota.decompiler.transformers.Transformer;
import moe.sota.decompiler.types.ClassType;
import moe.sota.decompiler.views.TabView;
import moe.sota.decompiler.views.TabsView;
import org.jetbrains.annotations.Nullable;
import org.koin.java.KoinJavaComponent;

import javax.swing.*;
import javax.swing.event.ChangeEvent;
import javax.swing.event.ChangeListener;
import java.awt.*;
import java.awt.event.ActionEvent;
import java.awt.event.ActionListener;

public class TabsController extends BaseController<TabsView> implements ActionListener, ChangeListener {

    @Getter
    public static final TabsController INSTANCE = new TabsController();

    private TabsController() {
        super(new TabsView());
        getView().getComboBox().addActionListener(this);
        getView().getModel().addChangeListener(this);
        String transformer = PreferenceService.PREFERENCES.get("transformer", null);
        if (transformer != null)
            getView().getComboBox().setSelectedItem(Transformer.valueOf(transformer));
    }

    @Override
    public void actionPerformed(ActionEvent event) {
        PreferenceService.PREFERENCES.put("transformer", getTransformer().name());
        for (int i = 0; i < getView().getTabCount(); i++) {
            TabController controller = ((TabView) getView().getComponentAt(i)).getController();
            if (controller.getFileModel().getType() instanceof ClassType)
                controller.updateAsync();
        }
    }

    @Override
    public void stateChanged(ChangeEvent changeEvent) {
        FileCloseTab fileCloseTab = KoinJavaComponent.get(FileCloseTab.class);
        fileCloseTab.setEnabled(getView().getTabCount() > 0);
    }

    public void addTab(FileModel fileModel) {
        TabController controller = getController(fileModel);
        if (controller == null) {
            controller = new TabController(fileModel);
            ImageIcon icon = new ImageIcon(fileModel.getIcon());
            Component component = controller.getComponent();
            controller.updateAsync().thenRun(() -> {
                if (getController(fileModel) == null) {
                    getView().addTab(fileModel.getName(), icon, component);
                    getView().setSelectedComponent(component);
                }
            });
        } else {
            getView().setSelectedComponent(controller.getComponent());
        }
    }

    public void closeTab() {
        getView().removeTabAt(getView().getSelectedIndex());
    }

    public void clearTabs() {
        getView().removeAll();
    }

    public Transformer getTransformer() {
        return (Transformer) getView().getComboBox().getSelectedItem();
    }

    private @Nullable TabController getController(FileModel fileModel) {
        TabController controller = null;
        for (int i = 0; i < getView().getTabCount(); i++) {
            TabController current = ((TabView) getView().getComponentAt(i)).getController();
            if (fileModel == current.getFileModel())
                controller = current;
        }
        return controller;
    }

}

File: src/main/kotlin/moe/sota/decompiler/controllers/TabsController.kt
package moe.sota.decompiler.controllers

import java.awt.event.ActionEvent
import java.awt.event.ActionListener
import javax.swing.event.ChangeEvent
import javax.swing.event.ChangeListener
import kotlinx.coroutines.MainScope
import kotlinx.coroutines.launch
import moe.sota.decompiler.menus.file.FileCloseTab
import moe.sota.decompiler.models.FileModel
import moe.sota.decompiler.services.PreferenceService
import moe.sota.decompiler.transformers.Transformer
import moe.sota.decompiler.types.ClassType
import moe.sota.decompiler.views.TabView
import moe.sota.decompiler.views.TabsView
import org.koin.java.KoinJavaComponent.get

class TabsController(
    private val tabsView: TabsView,
    private val createTabController: (FileModel) -> TabController,
) : ActionListener, ChangeListener {
    private val scope = MainScope()

    val transformer: Transformer?
        get() = tabsView.comboBox.selectedItem as Transformer?

    init {
        tabsView.comboBox.addActionListener(this)
        tabsView.getModel().addChangeListener(this)
        val transformer = PreferenceService.PREFERENCES.get("transformer", null)
        if (transformer != null) tabsView.comboBox.setSelectedItem(Transformer.valueOf(transformer))
    }

    override fun actionPerformed(event: ActionEvent?) {
        PreferenceService.PREFERENCES.put("transformer", this.transformer?.name)
        for (i in 0..<tabsView.tabCount) {
            val controller: TabController? = (tabsView.getComponentAt(i) as TabView).tabController
            if (controller?.fileModel?.type is ClassType) scope.launch { controller.update() }
        }
    }

    override fun stateChanged(changeEvent: ChangeEvent?) {
        val fileCloseTab = get<FileCloseTab>(FileCloseTab::class.java)
        fileCloseTab.setEnabled(tabsView.tabCount > 0)
    }

    fun addTab(fileModel: FileModel) {
        val existing = getController(fileModel)
        if (existing != null) {
            tabsView.setSelectedComponent(existing.tabView)
            return
        }

        val controller = createTabController(fileModel)
        val icon = fileModel.icon
        val component = controller.tabView
        scope.launch {
            controller.update()
            if (getController(fileModel) == null) {
                tabsView.addTab(fileModel.name, icon, component)
                tabsView.setSelectedComponent(component)
            }
        }
    }

    fun closeTab() {
        tabsView.removeTabAt(tabsView.selectedIndex)
    }

    fun clearTabs() {
        tabsView.removeAll()
    }

    private fun getController(fileModel: FileModel?): TabController? {
        var controller: TabController? = null

        for (i in 0..<tabsView.tabCount) {
            val current: TabController? = (tabsView.getComponentAt(i) as TabView).tabController
            if (fileModel === current?.fileModel) controller = current
        }

        return controller
    }
}
